### DISCUSSION TITLE
No compiler check on class argument types?

### DISCUSSION THREAD

**Discussion by Xylopyrographer** [espressif/arduino-esp32#9461] [Question - Community Help]
I was doing some testing on a library and came across puzzling result.

In the library `.h` file, the class constructor is defined as

```
class Button {
public:
    Button( uint8_t pin, uint32_t dbTime = 25, bool puEnable = true, bool invert = true ) {}
.
.
.
```

In the sketch `setup()` I was testing with a case where an object is declared as:

`Button myBtn( -76, 25, true, true );`

I expected this to fail as the first argument in the class constructor is of type `uint8_t` and in the test case `-76` is not of that type.

However it compiled without warning or error.

Wonder if anyone would have an explanation?

Using:
arduino-esp32 core 2.14
Arduino IDE Version: 2.3.2

Thanks.

**Comment by me-no-dev** [comment]
the compiler turned it into uint. It will compile, but would not run on 3.0.0, because the peripheral manager will say that the pin is invalid

**Reply by Xylopyrographer** [reply]
Is there a way to declare the constructor so such an error would be caught at compile time?

Didn't extend the test to see what would happen if this was attempted to be run on the device. Could be interesting.

**Reply by me-no-dev** [reply]
I don't think so. Not for such basic types. You could make an enum with all pins and accept that type as argument.
```cpp
typedef enum {
    PIN_0,
    PIN_1,
} my_pin_t;

class Button {
public:
	Button(my_pin_t pin);
};
```

**Reply by Xylopyrographer** [reply]
Appreciate the reply. Thanks.
